cpp/src/arrow/flight/serialization_internal.cc

```diff
@@ -272,7 +282,13 @@ grpc::Status FlightDataSerialize(const FlightPayload& msg, ByteBuffer* out,
         // entries are invalid.
         if (!buffer) continue;
 
-        slices.push_back(SliceFromBuffer(buffer));
+        grpc::Slice slice;
+        auto status = SliceFromBuffer(buffer).Value(&slice);
+        if (ARROW_PREDICT_FALSE(!status.ok())) {
+          // This will likely lead to abort as gRPC cannot recover from an error here
+          return ToGrpcStatus(status);
+        }
+        slices.push_back(slice);
```

Comment:
(not sure how `grpc::Slice` behaves wrt copies)

```suggestion
        slices.push_back(std::move(slice));
```